### PR TITLE
Iox #496 improve coverage for shared memory user

### PR DESCRIPTION
**Pull request by shankar-in**
## Pre-Review Checklist for the PR Author

1. [ ] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [ ] Tests follow the [best practice for testing][testing]
1. [ ] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [ ] Commits messages are according to this [guideline][commit-guidelines]
    - [ ] Commit messages have the issue ID (`iox-#123 commit text`)
    - [ ] Commit messages are signed (`git commit -s`)
    - [ ] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [ ] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [ ] Relevant issues are linked
1. [ ] Add sensible notes for the reviewer
1. [ ] All checks have passed (except `task-list-completed`)
1. [ ] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer
This PR is created for @manel94 Pls complete the above details.

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes **TBD**

### CODE REVIEW

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2021 by Robert Bosch GmbH. All rights reserved.
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:1]: Please use our clang-format.

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:3]: Please read the contribution file: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:7]: We are very thankful for every contribution but we require profound C++ knowledge as we stated before multiple times in our developer meetups. We also require that you please follow our coding guidelines.

- If you have architectural questions we are happy to help and provide you with the correct documentation.
- If you have a lack of C++ knowledge we can provide you with some book recommendations (see the last section of  https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md) so that you can refresh your knowledge before starting to contribute.

The reason for those requirements are that we develop safety critical software where maybe at some time in the future human lifes will depend on and we all want to make sure that we deliver the best possible software.

In this pull request we see a basic lack of knowledge of very basic C and C++ concepts. Furthermore, it seems that the basic functionality of iceoryx is not fully understood and the tests are actually testing nothing.

**Review by elfenpiff — CHANGES_REQUESTED** [review]
This book Clean C++ - Stephan Roth (https://www.amazon.de/-/en/Stephan-Roth/dp/1484227921) is a good starting point to get familiar with C++.

We would be happy to receive contributions from you in the future but for now we have to sadly decline them since we require a profound C++ knowledge.

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp (hunk 3 above)
Comment (budrus) [reply iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:7]: @elfenpiff I only see comments for the file header. What are the code parts that led to your conclusions? I think it would be fair to talk about specific things that do not follow the contribution guidelines or are wrong.

Comment (elfenpiff) [reply iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:7]: @budrus You are completely right, this is fair!

@manel94 sorry for that.

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    const uint64_t segmentId{1U};
+    rp::BaseRelativePointer::offset_t segmentManagerAddressOffset{0U};
+    void SetUp(){}
+    void TearDown(){}
+};
+
+TEST_F(SharedMemoryUserTest, ConstructorShmObjectWithZeroTopicSizeAndDoMapSharedMemoryIntoThreadBoolSetToTrueReturnError)
+{
+    iox::cxx::optional<iox::Error> detectedError;
+    const bool doMapSharedMemoryIntoThread{"TRUE"};
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:43]: Please never assign a string literal to a bool. This will result in a bool value which is always true. It is like writing `int a = "five";`

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    SharedMemoryUser ShmUser(doMapSharedMemoryIntoThread, topicSize, segmentId, segmentManagerAddressOffset);
+  
+    ASSERT_TRUE(detectedError.has_value());
+    EXPECT_EQ(detectedError.value(), iox::Error::kPOSH__SHM_APP_MAPP_ERR);
+}
+
+TEST_F(SharedMemoryUserTest, ConstructorShmObjectWithZeroTopicSizeAndDoMapSharedMemoryIntoThreadBoolSetToFalseReturnError)
+{
+    iox::cxx::optional<iox::Error> detectedError;
+    const bool doMapSharedMemoryIntoThread{"FALSE"};
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:60]: This boolean contains actually the value `true` since the string literal is, after it is casted to bool, not zero therefore true.

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+{
+  public:
+    RouDiEnvironment m_roudiEnv{iox::RouDiConfig_t().setDefaults()};
+    const uint64_t segmentId{1U};
+    rp::BaseRelativePointer::offset_t segmentManagerAddressOffset{0U};
+    void SetUp(){}
+    void TearDown(){}
+};
+
+TEST_F(SharedMemoryUserTest, ConstructorShmObjectWithZeroTopicSizeAndDoMapSharedMemoryIntoThreadBoolSetToTrueReturnError)
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:40]: This title describes the implementation and not what we actually do and expect. Maybe a title like `CreateAndMapZeroSizedSharedMemoryObjectFails` describes better what is actually done from a technical point of view.

Please adapt all the other titles as well.

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  
+    ASSERT_TRUE(detectedError.has_value());
+    EXPECT_EQ(detectedError.value(), iox::Error::kPOSH__SHM_APP_MAPP_ERR);
+}
+
+TEST_F(SharedMemoryUserTest, ConstructorShmObjectWithZeroTopicSizeAndDoMapSharedMemoryIntoThreadBoolSetToFalseReturnError)
+{
+    iox::cxx::optional<iox::Error> detectedError;
+    const bool doMapSharedMemoryIntoThread{"FALSE"};
+    const size_t topicSize{0U};
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:61]: Please remember the rules:
- zero
- one
- many
- limits
- corner cases

This means for the topic size:
- can we create a topic of size 0 (done already)
- can we create a topic of size 1 (done already)
- can we create a huge topic, maybe 1GB
- what happens when we create a topic which is larger then the available memory like for instance `2^64`

This means for the segmentId
- try to create multiple shared memory users with the same segment id should fail
- try to create multiple shared memory users with different segment ids should succeed
- try to create a large amount of tiny shared memory users with different segment ids should succeed

This means for the segmentManagerAddressOffset
- try to create multiple shared memory users with the same segmentManagerAddressOffset should fail
- try to create multiple shared memory users with the different segmentManagerAddressOffsets should succeed

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    });
+
+    SharedMemoryUser ShmUser(doMapSharedMemoryIntoThread, topicSize, segmentId, segmentManagerAddressOffset);
+    EXPECT_FALSE(detectedError.has_value());
+}
+
+TEST_F(SharedMemoryUserTest, ConstructorShmObjectWithValidTopicSizeAndDoMapSharedMemoryIntoThreadBoolSetToTrueReturnNoError)
+{
+    iox::cxx::optional<iox::Error> detectedError;
+    const bool doMapSharedMemoryIntoThread{"TRUE"};
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:93]: There are tests missing which verify that the shared memory was actually mapped into the process. The opposite can maybe tested with a death test.

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp (hunk 3 above)
Comment (budrus) [reply iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:7]: Thanks @elfenpiff

File: iceoryx_posh/test/moduletests/test_shared_memory_user.cpp (hunk 7 above)
Comment (manel94) [reply iceoryx_posh/test/moduletests/test_shared_memory_user.cpp:61]: @elfenpiff  after testing creating multiple shared memory users with the same segment id, same segmentManagerAddressOffset  don´t fail, is this a bug ?

**Comment by shankar-in** [comment]
@manel94  As requested by the author this PR will be declined.
